synapse/handlers/room.py

```diff
@@ -1450,25 +1451,37 @@ async def get_event_for_timestamp(
                         remote_response,
                     )
 
-                    # TODO: Do we want to persist this as an extremity?
-                    # TODO: I think ideally, we would try to backfill from
-                    # this event and run this whole
-                    # `get_event_for_timestamp` function again to make sure
-                    # they didn't give us an event from their gappy history.
                     remote_event_id = remote_response.event_id
                     origin_server_ts = remote_response.origin_server_ts
 
+                    # Backfill this event so we can get a pagination token for
+                    # it with `/context` and paginate `/messages` from this
+                    # point.
+                    #
+                    # FIXME: After this backfill, we might want to run this
+                    # whole `get_event_for_timestamp` function again to make
+                    # sure they didn't give us an event from their gappy
+                    # history. Also need a heuristic for when to stop recursing
+                    # if they keep giving us gappy results.
+                    await self.federation_event_handler.backfill_event(
+                        domain, room_id, remote_event_id
+                    )
```

Comment:
Main part of the PR is here. We now backfill the event that we fetch from the remote federated server.